kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterFactory.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.filter;
+
+import io.kroxylicious.proxy.InvalidConfigurationException;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * FilterFactory is a pluggable source of Kroxylicious filter implementations.
+ * @param <B> the configuration type for the Filter (use {@link Void} if the Filter is not configurable)
+ */
+public interface FilterFactory<B> {
+
+    /**
+     * The concrete type of the service this Contributor can instantiate
+     *
+     * @return type of the service this Contributor offers.
+     */
+    @NonNull
+    Class<? extends Filter> getServiceType();
+
+    /**
+     * The type of config expected by the service.
+     * <br/>
+     * The type must have a constructor annotated with the JsonCreator annotation.
+     * If the service has no configuration, return {@link Void} instead.
+     *
+     * @return type of config expected by the service.
+     */
+    @NonNull
+    Class<B> getConfigType();
+
+    /**
+     * Validate the configuration. By default, the configuration is considered invalid if
+     * the config type is not {@link Void} and the configuration is null.
+     * @param config configuration
+     * @throws InvalidConfigurationException when the configuration is invalid
+     */
+    default void validateConfiguration(B config) {
+        boolean requiresConfiguration = getConfigType() != Void.class;
+        if (requiresConfiguration && config == null) {
+            throw new InvalidConfigurationException(getServiceType().getSimpleName() + " requires configuration, but config object is null");
```

Comment:
What do you think of using an Exception? It is an exceptional case so makes sense from that point of view. We could pass back a set of failure message Strings or have some result object instead.

Reply:
I think an exception is probably the right move, but I wondered if it should offer a collection of validation errors? 

e.g. 
```
var invalidConfig =  new InvalidConfigurationException(getServiceType().getSimpleName() + " requires configuration, but config object is null");
invalidConfig.add("object.portNumber expected a valid port number in the ephemeral range but got 80");
....
if(invalidConfig != null) {
throw invalidConfig;
}
```

Reply:
Having to construct the exception conditionally might be a bit fiddly. Maybe it would be better to make the constructor take a set so they can go map from `config -> Set<String> -> InvalidFilterConfigurationException`
 
 So maybe the constructor should be like `new InvalidFilterConfigurationException(Set.of("failure1", "failure2"));`

Reply:
> So maybe the constructor should be like new InvalidFilterConfigurationException(Set.of("failure1", "failure2"));

+1 

Lazy construction would be optional and as you say a tricky but I'd be happy with the ctor accepting a set. Or maybe even `InvalidFilterConfigurationException(String... errors);`

Reply:
Also note this doesn't need to change in this PR!!

Reply:
I think we should defer it for now.